packages/connect/src/api/firmware/verifyEntropy.ts

```diff
@@ -0,0 +1,167 @@
+    for (let i = 0; i < a.length; i++) {
+        result[i] = a[i] ^ b[i];
+    }
+
+    return result;
+};
+
+// https://github.com/trezor/python-shamir-mnemonic/blob/master/shamir_mnemonic/cipher.py
+// simplified "decrypt" function
+const entropyToSeedSlip39 = async (encryptedSecret: Buffer) => {
+    const iterationExponent = 1;
+    // const identifier = 0;
+    // const extendable = true,
+    const passphrase = Buffer.from('', 'utf-8'); // empty passphrase
+    const salt = Buffer.alloc(0); // extendable: True => no salt
+
+    const half = Math.floor(encryptedSecret.length / 2);
+    let l = encryptedSecret.subarray(0, half);
+    let r = encryptedSecret.subarray(half);
+    for (let round = ROUND_COUNT - 1; round >= 0; round--) {
+        const f = await roundFunction(round, passphrase, iterationExponent, salt, r);
+        const rr = xor(l, f);
+        l = r;
+        r = rr;
+    }
+
+    return Buffer.concat([r, l]);
+};
+
+const getEntropy = (trezorEntropy: string, hostEntropy: string, strength: number) => {
+    const data = Buffer.concat([
+        Buffer.from(trezorEntropy, 'hex'),
+        Buffer.from(hostEntropy, 'hex'),
+    ]);
+    const entropy = sha256(data);
+
+    return Buffer.from(entropy.subarray(0, Math.floor(strength / 8)));
+};
+
+const computeSeed = (type: VerifyEntropyOptions['type'], secret: Buffer) => {
+    const BackupType = PROTO.Enum_BackupType;
+    if (
+        type &&
+        [
+            BackupType.Slip39_Basic,
+            BackupType.Slip39_Advanced,
+            BackupType.Slip39_Single_Extendable,
+            BackupType.Slip39_Basic_Extendable,
+            BackupType.Slip39_Advanced_Extendable,
+        ].includes(type)
+    ) {
+        // use slip39
+        return entropyToSeedSlip39(secret);
+    }
+
+    // use bip39
+    return mnemonicToSeed(entropyToMnemonic(secret));
+};
+
+const verifyCommitment = (entropy: string, commitment: string) => {
+    const hmacDigest = hmac(sha256, Buffer.from(entropy, 'hex'), Buffer.alloc(0));
+    if (!Buffer.from(hmacDigest).equals(Buffer.from(commitment, 'hex'))) {
+        throw new Error('Invalid entropy commitment');
+    }
+};
+
+type VerifyEntropyOptions = {
+    type?: PROTO.Enum_BackupType; // ResetDevice.backup_type
+    strength?: number; // ResetDevice.strength
+    commitment?: string; // entropy_commitment received from previous EntropyRequest
+    hostEntropy: string; // host_entropy used in previous EntropyAck
+    trezorEntropy?: string; // prev_entropy received from current EntropyRequest, after ResetDeviceContinue
+    xpubs: Record<string, string>; // <address_n, xpub>
+};
+
+export const verifyEntropy = async ({
+    type,
+    strength,
+    trezorEntropy,
+    hostEntropy,
+    commitment,
+    xpubs,
+}: VerifyEntropyOptions) => {
+    try {
+        if (!trezorEntropy || !commitment || !strength || Object.keys(xpubs).length < 1) {
+            throw new Error('Missing verifyEntropy data');
```

Comment:
When an error is thrown here, Suite resets to the default screen but device hangs on loading screen until the user interacts with it again via Suite.

![Screenshot 2025-01-03 at 16 52 37](https://github.com/user-attachments/assets/f63b4ac6-d9b4-4ff3-82fe-18306ed23464)

Reply:
fixed, `resetDevice` will send `Cancel` message before throwing `Failure_EntropyCheck`